### PR TITLE
Add Web api on Owin approach to LiteRestApp.

### PR DESCRIPTION
**Pull request by chenghuang-mdsol**
To choose a listener (e.g. web api listener), use LiteRestApp.Use(IListen)

### CODE REVIEW

**Comment by junshao** [comment]
@chenghuang-mdsol Medidata has a convention that a developer shouldn't merge the PR opened by himself. A PR should be merged by another developer. We follow this rule to ensure every change can be aware by at least two developers.

File: Medidata.Cloud.Thermometer/Controllers/ThermometerController.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using System.Web.Http;
+using System.Web.Http.Results;
+using Medidata.Cloud.Thermometer.WebApiOnOwin;
+using Newtonsoft.Json;
+
+namespace Medidata.Cloud.Thermometer.Controllers
+{
+    public class ThermometerController : ApiController
```
Comment (junshao) [inline Medidata.Cloud.Thermometer/Controllers/ThermometerController.cs]: This controller should be in the caller assembly but not in the Thermometer itself.

Comment (junshao) [inline Medidata.Cloud.Thermometer/Controllers/ThermometerController.cs]: You can move this class to Medidata.Cloud.Thermometer.Example project.

**Comment by junshao** [comment]
@chenghuang-mdsol Though still under discussion, I think WebApi probably won't be our first choice. I have a comment explaining my concerns on https://mdsol.jiveon.com/community/research-development/app-dev/study-conduct-team/rave/blog/2015/07/27/diagnostic-api-technologies#comment-8522. 

However I am very surprised that you made this PR with many concrete trails so soon. :+1: Keep going.

File: Medidata.Cloud.Thermometer/Listeners/WebApiOnOwinListener.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+using Owin;
+
+namespace Medidata.Cloud.Thermometer.Listeners
+{
+    public class WebApiOnOwinListener : IListen
+    {
+        public Task<IDisposable> Listen(int port)
+        {
+
+            if (port <= 1024) throw new ArgumentException("Must choose a port greater than 1024", "port");
```
Comment (junshao) [inline Medidata.Cloud.Thermometer/Listeners/WebApiOnOwinListener.cs]: Maybe checking `<=1024` is overengineering.

File: Medidata.Cloud.Thermometer/Listeners/WebApiOnOwinListener.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+                //    defaults: new { id = RouteParameter.Optional }
+                //    );
+                config.EnsureInitialized();
+                builder.UseWebApi(config);
+
+            });
+        }
+
+        //Web Api actually doesn't use the Action Route, the actions are in the "Controllers"
+        public ActionRouteConfiguration Routes { get; private set; }
```
Comment (junshao) [inline Medidata.Cloud.Thermometer/Listeners/WebApiOnOwinListener.cs]: If don't use, I think we can throw a `NotSupportedException` here.

File: Medidata.Cloud.Thermometer/Medidata.Cloud.Thermometer.csproj [hunk 4]
```diff
@@ -33,6 +33,10 @@
     <RunPostBuildEvent>Always</RunPostBuildEvent>
   </PropertyGroup>
   <ItemGroup>
+    <Reference Include="Autofac">
+      <HintPath>..\packages\Autofac.3.5.0\lib\net40\Autofac.dll</HintPath>
+      <Private>True</Private>
+    </Reference>
```
Comment (junshao) [inline Medidata.Cloud.Thermometer/Medidata.Cloud.Thermometer.csproj]: Nice IoC consideration. :+1:

**Comment by junshao** [comment]
I think I see your idea.
> Implement using different framework in different namespace and use IoC to switch from different implementatio

Is it your idea? Very nice.

What if we only define interface but implement in different NuGet packages? Like
- Medidata.Rave.Thermometer NuGet package, which provide basic pure Owin implementation.
- Medidata.Rave.Thermometer.WebApiListener NuGet package, which provide WebAPI implementation.
- Medidata.Rave.Thermometer.Nancy NuGet package
- Medidata.Rave.Thermometer.LightNode NuGet package

And usage can be like
```cs
// Pure Owin
new ThermoApp().WhenGet(...).WhenGet(...).Listen(9000); 

// If using WebApi
new ThermoApp(new WebApiListener()).WhenGet(...).WhenGet(...).Listen(9000);

// If using Nancy
new ThermoApp(new NancyListener()).WhenGet(...).WhenGet(...).Listen(9000);

// If using LightNode
new ThermoApp(new LightNodeListener()).WhenGet(...).WhenGet(...).Listen(9000);
```

**Comment by junshao** [comment]
In real code, we can also use IoC to control the listener in the caller assembly as well.

**Comment by junshao** [comment]
But probably it's hard to design a ubiquitous usage that can fulfill all these three frameworks. Like
```cs
class RaveImplController: ApiControllerAndNancyControllerAndLightNodeController {
  ...
}
```

File: Medidata.Cloud.Thermometer/Listeners/WebApiOnOwinListener.cs (hunk 3 above)
Comment (chenghuang-mdsol) [inline Medidata.Cloud.Thermometer/Listeners/WebApiOnOwinListener.cs]: That makes sense

File: Medidata.Cloud.Thermometer/Controllers/ThermometerController.cs (hunk 1 above)
Comment (chenghuang-mdsol) [inline Medidata.Cloud.Thermometer/Controllers/ThermometerController.cs]: Got it. Do you want me to revert it?

**Comment by chenghuang-mdsol** [comment]
I tried to separate them but stopped by some circular dependency issue.
I can try again after I finish the unit tests.

**Comment by junshao** [comment]
I suggest you adding UT when the design is relatively stable. Otherwise, it'll double our workload if we changed our minds.

File: Medidata.Cloud.Thermometer/Controllers/ThermometerController.cs (hunk 1 above)
Comment (junshao) [inline Medidata.Cloud.Thermometer/Controllers/ThermometerController.cs]: We can keep it. Anyway this repo is just a playground only for us for now. No any harm to other people.

**Pull request merged**